File: 2022/day-10/day10.py

import dataclasses
import io
from typing import *
from functools import lru_cache, partial


@dataclasses.dataclass
class CPU:
    clock: int
    register: int


Instruction = Callable[[CPU], Iterable]
Program = Iterable[Instruction]


def noop(_: CPU):
    yield


def addx(v: int, cpu: CPU):
    yield from noop(cpu)
    cpu.register += v
    yield


def parse_input(inp: str) -> Program:
    @lru_cache(128)
    def get_command(l: str):
        if l.startswith('noop'):
            return noop
        elif l.startswith('addx'):
            _, v = l.split(' ')
            return partial(addx, int(v))
        else:
            raise ArithmeticError
    return map(get_command, inp.splitlines())


def run(program: Program, cpu: Optional[CPU] = None, offset: int = 20, interval: int = 40) -> Iterable[int]:
    cpu = cpu or CPU(1, 1)
    display = io.StringIO()
    for instruction in program:
        for _ in instruction(cpu):
            print_pixel(cpu, file=display)
            cpu.clock += 1
            if (cpu.clock - offset) % interval == 0:
                yield cpu.clock * cpu.register

    display = display.getvalue()
    for line in display.splitlines():
        print(line[-1:] + line[:-1])  # rotate line by one


def print_pixel(cpu: CPU, width: int = 40, file=None):
    print('#' if abs(cpu.clock % width - cpu.register) <= 1 else '.', end='', file=file)
    if cpu.clock % width == 0:
        print(file=file)


with open('input.txt') as h:
    inp = h.read()
print('Part 2:')
print('Part 1:', sum(run(parse_input(inp))))

File: 2022/day-10/day10.rs

fn run_program(program: &str) {
    let mut register = 1_i32;
    let mut register_history: Vec<i32> = Vec::new();

    program.lines().for_each(|l| {
        if l.starts_with("noop") {
            register_history.push(register);
        } else if l.starts_with("addx") {
            let diff = l
                .split(' ')
                .take(2)
                .last()
                .unwrap()
                .parse::<i32>()
                .expect("Failed to parse number");
            register_history.push(register);
            register_history.push(register);
            register += diff;
        }
    });

    let signal_strength: i32 = register_history
        .iter()
        .enumerate()
        .skip(19)
        .step_by(40)
        .map(|(clock, &reg)| ((clock + 1) as i32) * reg)
        .sum();
    println!("Part 1: {:?}", signal_strength);

    let display_width = 40_usize;
    let display: Vec<char> = register_history
        .iter()
        .enumerate()
        .map(|(i, &reg)| match (i % display_width) as i32 - reg {
            (-1..=1) => '#',
            _ => '.',
        })
        .collect();

    println!("Part: 2");
    for line in display[..].chunks(40) {
        println!("{}", line.iter().collect::<String>());
    }
}

fn main() {
    let contents = std::fs::read_to_string("input.txt").expect("Failed to read input");
    run_program(&contents);
}
